### DISCUSSION TITLE
Detecting Windows App SDK version at runtime

### DISCUSSION THREAD

**Discussion by FreddyDgh** [microsoft/WindowsAppSDK#2796] [Q&A] [answered]
If I have a Windows App SDK 1.1 app, is there a reliable way to determine which version of the runtime is being used?  Obviously, it will be 1.1.x release, but I'd like to know precisely if it is 1.1.1 or 1.1.2 or 1.1.3 etc.  I can't seem to find an obvious way of doing this, and it would be very helpful for analytics/bug-tracking purposes.

**Comment by castorix** [comment]
You can see this similar question : [Recommended way to check that the Windows App SDK is installed](https://github.com/microsoft/WindowsAppSDK/discussions/2437)

**Reply by FreddyDgh** [reply]
@castorix I think that approach could work, but the code in that discussion uses `global::Microsoft.WindowsAppSDK.Runtime.Packages.Framework.PackageFamilyName`, which doesn't seem to exist in version 1.1.  Has the functionality been moved somewhere else?

**Reply by FreddyDgh** [reply]
I have figured out the issues with the constants:

For C++, they are in `WindowsAppSDK-VersionInfo.h`, documented at https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/win32/_bootstrap/

For C#, they are in `WindowsAppSDK-VersionInfo.cs`, documented at https://docs.microsoft.com/en-us/windows/apps/api-reference/cs-bootstrapper-apis/cs-bootstrapper-apis-constants.  However, the docs outright lie and say that the constants are in the assembly `Microsoft.WindowsAppRuntime.Bootstrap.Net.dll` when, in fact, they are included in the output assembly when you use `<WindowsPackageType>None</WindowsPackageType>`.  It appears that if you don't want the default bootstrapping behavior, your only real option for a workaround is to change the project file like so:

```
	<ItemGroup>
		<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.1.3" GeneratePathProperty="true" />
	</ItemGroup>

	<ItemGroup>
		<Compile Include="$(PkgMicrosoft_WindowsAppSDK)\include\WindowsAppSDK-VersionInfo.cs" />
	</ItemGroup>
```

@DrusTheAxe The C# workaround is pretty clunky and non-obvious.  Is it possible to fix this?  Seems strange that those constants are not included directly in the assembly in the first place; it seems like a much more logical place for them.

**Comment by DrusTheAxe** [comment, marked as answer]
> If I have a Windows App SDK 1.1 app, is there a reliable way to determine which version of the runtime is being used?

Yes. Do you want the version of the MSIX packages? Or the version of the WinAppSDK release (1.1.3, etc)?

For the former, walk the current process' package graph looking for the 1st match for the framework package's family name, then the package's version field has the answer.

For the latter do same, but then pull out the major version's bottom 3 digits for the patch level e.g. WinAppSDK release 1.1.3 has MSIX packages with version `1003.x.y.z`. You already know the release's major.minor version you're built with (the '1.1' in '1.1.3') so the MSIX package `version.major % 1000` gets you the patch level (the '3' in '1.1.3').

[WindowsAppRuntime_IsSelfContained()](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/Common/WindowsAppRuntime.SelfContained.cpp) does this. It's C++ (not C#) but implements this algorithm.

>the docs outright lie and say that the constants are in the assembly Microsoft.WindowsAppRuntime.Bootstrap.Net.dll when, in fact, they are included in the output assembly when you use <WindowsPackageType>None</WindowsPackageType>.

Hmmm. The phrasing may be confusing. The constants via `Microsoft.WindowsAppRuntime.Bootstrap.Net.dll` are compile-time values you're  built with embedded into the caller, as I understand it. This is the moral equivalent of `WindowsAppSDK-VersionInfo.h`, or at least meant to be. Sounds like this wasn't clearly conveyed?

I'm working on a _runtime_ API to complement this build-time API, equivalent-ish to what's exposed by WindowsAppSDK-VersionInfo.* but functions instead of constants. Thus a runtime lookup retrieving values _from the runtime binaries currently in use_, not what you built with. I hope to have something concrete to share shortly.

**Reply by FreddyDgh** [reply]
> I'm working on a _runtime_ API to complement this build-time API, equivalent-ish to what's exposed by WindowsAppSDK-VersionInfo.* but functions instead of constants. Thus a runtime lookup retrieving values _from the runtime binaries currently in use_, not what you built with. I hope to have something concrete to share shortly.

This sounds like exactly what I would want.  The SDK version that the app is built with is nice to have as well, but not as critical as the version of binaries currently in use.  I am hoping that the planned API will NOT use the same approach as `WindowsAppSDK-VersionInfo.cs` where it includes an extra .cs file in the final project.

**Reply by Balkoth** [reply]
Is this
```
RETURN_IF_FAILED(::AppModel::PackageGraph::GetCurrentPackageGraph(flags, packageInfoCount, packageInfo, buffer));
const auto frameworkPackageFamilyName{ ::WindowsAppRuntime::VersionInfo::Framework::GetPackageFamilyName() };
```
somewhere exposed in .net?

**Reply by castorix** [reply]
> Is this
> 
> ```
> RETURN_IF_FAILED(::AppModel::PackageGraph::GetCurrentPackageGraph(flags, packageInfoCount, packageInfo, buffer));
> const auto frameworkPackageFamilyName{ ::WindowsAppRuntime::VersionInfo::Framework::GetPackageFamilyName() };
> ```
> 
> somewhere exposed in .net?

You just call [GetCurrentPackageInfo ](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentpackageinfo)with P/Invoke

**Reply by Balkoth** [reply]
This does not work for unpackaged apps as `GetCurrentPackageInfo` returns `APPMODEL_ERROR_NO_PACKAGE`.

**Reply by DrusTheAxe** [reply]
> This does not work for unpackaged apps as `GetCurrentPackageInfo` returns `APPMODEL_ERROR_NO_PACKAGE`.

It does if the process has no package identity (aka unpackaged app) AND the package graph is empty (for compat reasons).

If you're an unpackaged app you can call [GetCurrentPackageInfo()](https://docs.microsoft.com/windows/win32/api/appmodel/nf-appmodel-getcurrentpackageinfo) but treat APPMODEL_ERROR_NO_PACKAGE as Success and the returned count=0.

To answer the earlier question: Yes, GetCurrentPackageInfo*() is _the_ API to get the package graph.

There's no WinRT or C#-specific equivalent today for unpackaged apps. Is that something you'd like to see? If so please open an Issue to get it onto our radar for consideration for a future enhancement. Thanks!

**Reply by DrusTheAxe** [reply]
> I'm working on a _runtime_ API to complement this build-time API

BTW still in the queue just delayed behind some other work. I'm hoping to have code and spec up next week.

**Reply by DrusTheAxe** [reply]
FYI PR for the proposed API @ [Runtime query'able version information #2889](https://github.com/microsoft/WindowsAppSDK/pull/2889)

Spec for the API coming shortly

**Reply by DrusTheAxe** [reply]
[Runtime query'able version information API spec #2893](https://github.com/microsoft/WindowsAppSDK/pull/2893)

**Reply by FreddyDgh** [reply]
@DrusTheAxe This API is exactly what I was looking for in my original scenario.  However, I am a little confused about the difference/relationship between `ReleaseInfo` and `RuntimeInfo` - if I understand correctly, both are resolved at runtime, but one is in the format "1.1.4" and the other is the actual version of the packages, e.g. "1004.584.2120.0".  Previously, I had thought the two were correlated in a one-to-one manner - is it possible for the them to change independently? For example, when it comes to logging/tracking bugs, it is it sufficient to track just the `ReleaseInfo` version or do I actually need to track both `ReleaseInfo` and `RuntimeInfo`?  Maybe this distinction could be clearer in the final documentation.

**Reply by DrusTheAxe** [reply]
> This API is exactly what I was looking for in my original scenario. 

:-)

>However, I am a little confused about the difference/relationship between `ReleaseInfo` and `RuntimeInfo` - if I understand correctly, both are resolved at runtime, but one is in the format "1.1.4" and the other is the actual version of the packages, e.g. "1004.584.2120.0". 

Correct

* 'Release' is the project version. Syntactically it's "major.minor-channel[N]" or "major.minor.patch" for the Stable channel e.g. 1.2-experimental1 vs 1.1.4. I guess to be pedantic that's "major.minor.patch-channel[N]" (e.g. 1.2.0-experimental1) but 'patch' is always zero for non-Stable channels so I always see it w/o the 'patch' (0).
* 'Runtime' is the MSIX package version. A DotQuadNumber so technically it's uint64 or uint16[4], denoted as "major.minor.build.revision"


>I had thought the two were correlated in a one-to-one manner

They are, but release->runtime can't be computed and runtime->release is doable but cumbersome and poses future compat issues.

_The longer story..._

MSIX version contains the can't be algorithmically derived from release version because it includes the date/time when the build runs e.g. 1004.584.2120.0 = major.1.004 i.e. major.1.4, built 584 days after Jan'1, 2020, at 2120 i.e. 9:20pm (I don't recall if UTC or local time and if the latter, which time zone). Thus ComputeRuntimeVersion(in releaseversion) can't be implemented.

...Release version can be computed from MSIX (runtime) version and the PackageFamilyName but not easily. Requires string parsing the family name for the release version's major.minor and version tag (channel + # e.g. preview2) and parsing the runtime's "major" field for the release's minor.patch. The latter's trivial and the former's "just" a bunch of string parsing so you _could_ write the code to do it. But it's irksome, and also becomes a data contract so if/when we change in the future poses compatibility issues. For instance, if we support 'runtime minor updates are drop in compatible' (e.g. 2.7.* can be used at runtime by apps built with 2.1) then the framework's package family name would be the same across minor releases e.g. "Microsoft.WindowsAppRuntime.2_8wekyb3d8bbwe". Parsers looking for major.minor would fail.

And even if it wasn't fragile, one of the points of WinAppSDK is to make devs lives easier. Sure you _could_ write parsing logic, but why make _you_ do it when _we_ can easily (and more efficiently) shrink wrap it for you? Throw in avoiding the fragility of a data contract and future seamless compat and the question isn't "Why provide both?" but rather, "Why not?" :-)

>is it possible for the them to change independently?

No.

>it is it sufficient to track just [one or the other?]

Depends what you intend to do with the data. If you're logging for support purposes `ReleaseInfo.Current().ToString()` might suffice e.g. "1.2-experimental1" or "1.1.4". If you're looking to do programmatic things with the data you probably need the runtime value.

>Maybe this distinction could be clearer in the final documentation.

Good input! Will keep in mind when we write the docs. And if you find the docs unclear or otherwise suboptimal, please share your feedback. We're always interested to hear if/when folks find docs less than helpful.

**Reply by Balkoth** [reply]
> > This does not work for unpackaged apps as `GetCurrentPackageInfo` returns `APPMODEL_ERROR_NO_PACKAGE`.
> 
> It does if the process has no package identity (aka unpackaged app) AND the package graph is empty (for compat reasons).
> 
> If you're an unpackaged app you can call [GetCurrentPackageInfo()](https://docs.microsoft.com/windows/win32/api/appmodel/nf-appmodel-getcurrentpackageinfo) but treat APPMODEL_ERROR_NO_PACKAGE as Success and the returned count=0.
> 
> To answer the earlier question: Yes, GetCurrentPackageInfo*() is _the_ API to get the package graph.
> 
> There's no WinRT or C#-specific equivalent today for unpackaged apps. Is that something you'd like to see? If so please open an Issue to get it onto our radar for consideration for a future enhancement. Thanks!

I do not understand how this is supposed to get a package graph in unpackaged apps. You say that returned count is zero, so there are zero [PACKAGE_INFO](https://docs.microsoft.com/en-us/windows/desktop/api/appmodel/ns-appmodel-package_info) structures in the buffer, so no package information is returned. Am i missing something?

Does the API spec you mentioned support unpackaged apps?

**Reply by DrusTheAxe** [reply]
>Does the API spec you mentioned support unpackaged apps?

Yes. That's the driving force why Dynamic Dependencies were invented - so unpackaged apps can use packaged bits.

All processes have a package graph, and have always had one. The difference is before Dynamic Dependencies (1) the package graph could only set set at process creation and (2) retrieving the package graph for unpackaged processes wasn't supported (and thus `APPMODEL_ERROR_NO_PACKAGE`).

If we had to do over again a "GetPackageGraph()" API would always succeed and right after process creation would return count=0 for an unpackaged app and 1+ for packaged apps. But `GetCurrentPackageInfo` was created years ago (Win8) and some folks rely on it failing with `APPMODEL_ERROR_NO_PACKAGE` so _for compatibilty reasons_ we added DynDep and taught GetCurrentPackageInfo about it, but didn't (couldn't) change returning that error if count=0.

>You say that returned count is zero, so there are zero [PACKAGE_INFO](https://docs.microsoft.com/en-us/windows/desktop/api/appmodel/ns-appmodel-package_info) structures in the buffer, so no package information is returned

Correct.

You'll never encounter that if you're a packaged app i.e. born with pkggraph=[].

--OR--

You're an unpackaged app and called the Bootstrap API (which adds WinAppSDK to the pkggraph).

The only way `GetCurrentPackageInfo` can reutnr APPMODEL_ERROR_NO_PACKAGE is if (1) you're an unpackaged app and (2) call `GetCurrentPackageInfo` before calling the Bootstrap API. The only scenario for this is if an unpackaged app uses WinAppSDK/SelfContained. Then the app itself isn't packaged AND doesn't call the Bootstrap API, thus package graph is empty.

**Reply by Balkoth** [reply]
What luck i am in, my app uses WinAppSDK/SelfContained...

**Reply by DrusTheAxe** [reply]
Are you an unpackaged app? If so and you're using WinAppSDK/SelfContained why would you call `GetCurrentPackageInfo`?

**Reply by Balkoth** [reply]
To get version info about the WinAppSDK in use as suggested above.

**Reply by DrusTheAxe** [reply]
If you're using WinAppSDK/SelfContained then the version you built with is exactly what you're using.

The new [VersionInfo API](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/VersionInfo/VersionInfo.md) doesn't probe the PackageGraph looking for the runtime so it works for WinAppSDK/SelfContained as well as WinAppSDK/MSIX (aka framework-dependent). 

This is going [into main shortly](https://github.com/microsoft/WindowsAppSDK/pull/2889) for 1.2

**Reply by FreddyDgh** [reply]
@DrusTheAxe Looks like the VersionInfo API didn't make the cut for 1.2 stable?  Do you know if we can expect it in a 1.2 servicing release, or will it not be stable until 1.3?

**Reply by fabianoriccardi** [reply]
This discussion is pretty old and marked as solved but I don't think it is. [Here](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/VersionInfo/VersionInfo.md) there are the specs about nice APIs to retrieve versions that are not yet implemented in v1.6

Moreover, it seems that part of this APIs are already available in win32/c++... [docs](
https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/windowsappsdk-versioninfo/ns-windowsappsdk-versioninfo-versioninfo).

I could be wrong, and these APIs are already available. In case, please let me know which is the one-liner (or a kind of) to retrieve the Windows App SDK version in C#.
